Author: Yasaman Monazzah

Date: 05/14/2024

Capstone Project

__________________________________________________________________________________________________________________________________________

## Table of Contents

Introduction

[1. - Initial Setup and Data Preparation](#part-1)


[2. - Overview of the Dataset](#part-2)


[3. - Data Cleaning and Preprocessing](#part-3)


[4. - Exploratory Data Analysis (EDA)](#part-4)


[5. - Next Steps: Data Processing, Feature Engineering, and Baseline Modeling](#part-5)

__________________________________________________________________________________________________________________________________________

## Introduction




Banks and financial organizations must determine which applicants are eligible for loans and which are not. The goal of this case study is to find patterns in the data to assess whether or not a specific person should be awarded a loan. This is especially important in the case that the applicant has no credit history and the application is declined. By recognizing trends in the data collection, financial institutions can make informed judgments and reduce business losses in the event of a loan default. This is because lending money to someone who is unlikely to repay the loan would result in financial losses for the institution. Understanding the characteristics that cause default is critical for businesses. Thus, it is vital to uncover patterns for clients who are unable to pay the loan by refusing the loan, lowering the loan amount, or raising the interest rate.

## 1. Initial Setup and Data Preparation
<a class="anchor" id="part-1"></a>

###  Setting Up Jupyter Notebook and Importing Libraries
<a class="anchor" id="part-1"></a>

In [1]:
# Importing data analysis libraries
import numpy as np
import pandas as pd
import plotly.express as px
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import seaborn as sns
import statsmodels.api as sm
import warnings
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from scipy import stats
from scipy.stats import pearsonr
from itertools import product
from IPython.display import display
%matplotlib inline

###  Setting Up Display Options in Jupyter Notebook

In [2]:
# Set display options to show maximum rows and columns
pd.set_option('display.max_rows', None)     # Display all rows
pd.set_option('display.max_columns', None)  # Display all columns
pd.options.display.float_format = '{:.2f}'.format

###  Handling Warnings

In [3]:
# Suppress specific FutureWarnings
warnings.filterwarnings("ignore", category=FutureWarning, message=".*use_inf_as_na.*")


### Importing Data Files into Jupyter Notebook

In [4]:
## Reading the CSV file into a pandas DataFrame
columns_description = pd.read_csv('columns_description.csv')

In [5]:
# Display the DataFrame in a tabular format
display(columns_description)

Unnamed: 0                     Table                           Row  \
0             1          application_data                    SK_ID_CURR   
1             2          application_data                        TARGET   
2             5          application_data            NAME_CONTRACT_TYPE   
3             6          application_data                   CODE_GENDER   
4             7          application_data                  FLAG_OWN_CAR   
5             8          application_data               FLAG_OWN_REALTY   
6             9          application_data                  CNT_CHILDREN   
7            10          application_data              AMT_INCOME_TOTAL   
8            11          application_data                    AMT_CREDIT   
9            12          application_data                   AMT_ANNUITY   
10           13          application_data               AMT_GOODS_PRICE   
11           14          application_data               NAME_TYPE_SUITE   
12           15          application_data              NAME_INCOME_TYPE   
13           16          application_data           NAME_EDUCATION_TYPE   
14           17          application_data            NAME_FAMILY_STATUS   
15           18          application_data             NAME_HOUSING_TYPE   
16           19          application_data    REGION_POPULATION_RELATIVE   
17           20          application_data                    DAYS_BIRTH   
18           21          application_data                 DAYS_EMPLOYED   
19           22          application_data             DAYS_REGISTRATION   
20           23          application_data               DAYS_ID_PUBLISH   
21           24          application_data                   OWN_CAR_AGE   
22           25          application_data                    FLAG_MOBIL   
23           26          application_data                FLAG_EMP_PHONE   
24           27          application_data               FLAG_WORK_PHONE   
25           28          application_data              FLAG_CONT_MOBILE   
26           29          application_data                    FLAG_PHONE   
27           30          application_data                    FLAG_EMAIL   
28           31          application_data               OCCUPATION_TYPE   
29           32          application_data               CNT_FAM_MEMBERS   
30           33          application_data          REGION_RATING_CLIENT   
31           34          application_data   REGION_RATING_CLIENT_W_CITY   
32           35          application_data    WEEKDAY_APPR_PROCESS_START   
33           36          application_data       HOUR_APPR_PROCESS_START   
34           37          application_data    REG_REGION_NOT_LIVE_REGION   
35           38          application_data    REG_REGION_NOT_WORK_REGION   
36           39          application_data   LIVE_REGION_NOT_WORK_REGION   
37           40          application_data        REG_CITY_NOT_LIVE_CITY   
38           41          application_data        REG_CITY_NOT_WORK_CITY   
39           42          application_data       LIVE_CITY_NOT_WORK_CITY   
40           43          application_data             ORGANIZATION_TYPE   
41           44          application_data                  EXT_SOURCE_1   
42           45          application_data                  EXT_SOURCE_2   
43           46          application_data                  EXT_SOURCE_3   
44           47          application_data                APARTMENTS_AVG   
45           48          application_data              BASEMENTAREA_AVG   
46           49          application_data   YEARS_BEGINEXPLUATATION_AVG   
47           50          application_data               YEARS_BUILD_AVG   
48           51          application_data                COMMONAREA_AVG   
49           52          application_data                 ELEVATORS_AVG   
50           53          application_data                 ENTRANCES_AVG   
51           54          application_data                 FLOORSMAX_AVG   
52           55          appli

In [6]:
## Reading the CSV file into a pandas DataFrame
df = pd.read_csv('application_data.csv')

In [7]:
# Convert inf values to NaN in the DataFrame
df.replace([np.inf, -np.inf], np.nan, inplace=True)

### 2. Overview of the Dataset
<a class="anchor" id="part-2"></a>

In [8]:
# Look at the first 5 rows
df.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0         202500.00   406597.50     24700.50   
1               N             0         270000.00  1293502.50     35698.50   
2               Y             0          67500.00   135000.00      6750.00   
3               Y             0         135000.00   312682.50     29686.50   
4               Y             0         121500.00   513000.00     21865.50   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0        351000.00   Unaccompanied          Working   
1       1129500.00          Family    State servant   
2        135000.00   Unaccompanied          Working   
3        297000.00   Unaccompanied          Working   
4        513000.00   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                        0.02       -9461           -637           -3648.00   
1                        0.00      -16765          -1188           -1186.00   
2                        0.01      -19046           -225           -4260.00   
3                        0.01      -19005          -3039           -9833.00   
4                        0.03      -19932          -3038           -4311.00   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531        26.00           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers             1.00   
1                 1           1           0      Core staff             2.00   
2                 1           1           0        Laborers             1.00   
3                 1           0           0        Laborers             2.00   
4                 1           0           0      Core staff             1.00   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     2                            2   
3                     2                            2   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
0                  WEDNESDAY                       10   
1                     MONDAY                       11   
2                     MONDAY                        9   
3                  WEDNESDAY                       17   
4                   THURSDAY                       11   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  \
0                           0                           0   
1                           0                           0   
2                        

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


In [10]:
# Look at 5 random rows
df.sample(5)

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
23935       127841       0         Cash loans           F            N   
104472      221241       0         Cash loans           F            N   
153874      278336       0         Cash loans           F            N   
102385      218859       0         Cash loans           F            N   
79458       192113       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
23935                Y             0         247500.00   762453.00   
104472               Y             0         135000.00   733315.50   
153874               Y             0         202500.00   675000.00   
102385               Y             0          67500.00   344043.00   
79458                Y             0         202500.00   315000.00   

        AMT_ANNUITY  AMT_GOODS_PRICE  NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
23935      80194.50        706500.00    Unaccompanied             Pensioner   
104472     41076.00        679500.00    Unaccompanied  Commercial associate   
153874     26284.50        675000.00  Spouse, partner               Working   
102385     14706.00        297000.00    Unaccompanied             Pensioner   
79458      17217.00        315000.00    Unaccompanied               Working   

                  NAME_EDUCATION_TYPE NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
23935                Higher education            Married  House / apartment   
104472  Secondary / secondary special            Married  House / apartment   
153874  Secondary / secondary special     Civil marriage  House / apartment   
102385  Secondary / secondary special     Civil marriage  House / apartment   
79458   Secondary / secondary special            Married  House / apartment   

        REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  \
23935                         0.05      -21036         365243   
104472                        0.05      -21767          -5390   
153874                        0.03      -13517           -174   
102385                        0.01      -21709         365243   
79458                         0.03      -16810          -9568   

        DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  \
23935             -489.00            -3487          NaN           1   
104472          -12996.00            -5094          NaN           1   
153874           -7542.00            -2949          NaN           1   
102385           -5959.00            -4533          NaN           1   
79458            -8303.00             -362          NaN           1   

        FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  FLAG_PHONE  \
23935                0                0                 1           0   
104472               1                0                 1           0   
153874               1                0                 1           0   
102385               0                0                 1           1   
79458                1                0                 1           1   

        FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  REGION_RATING_CLIENT  \
23935            0             NaN             2.00                     1   
104472           0        Managers             2.00                     1   
153874           0        Laborers             2.00                     2   
102385           0             NaN             2.00                     2   
79458            0        Laborers             2.00                     2   

        REGION_RATING_CLIENT_W_CITY WEEKDAY_APPR_PROCESS_START  \
23935                             1                  WEDNESDAY   
104472                            1                     MONDAY   
153874                            2                     FRIDAY   
102385                            2                   SATURDAY   
79458                             2                   THURSDAY   

        HOUR_APPR_PROCESS_START  REG_REGION_NOT_LIVE_REGION  \
23935       

In [11]:
# Check the dimensions of the dataset
df.shape

(307511, 122)

In [12]:
print(f'There are {df.shape[0]} rows and {df.shape[1]} columns in this dataset.')

There are 307511 rows and 122 columns in this dataset.


In [13]:
#Get an overview of dataset variables
#When verbose=True, it provides more detailed information about the DataFrame, including the data types of each column and the memory usage
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Data columns (total 122 columns):
 #    Column                        Dtype  
---   ------                        -----  
 0    SK_ID_CURR                    int64  
 1    TARGET                        int64  
 2    NAME_CONTRACT_TYPE            object 
 3    CODE_GENDER                   object 
 4    FLAG_OWN_CAR                  object 
 5    FLAG_OWN_REALTY               object 
 6    CNT_CHILDREN                  int64  
 7    AMT_INCOME_TOTAL              float64
 8    AMT_CREDIT                    float64
 9    AMT_ANNUITY                   float64
 10   AMT_GOODS_PRICE               float64
 11   NAME_TYPE_SUITE               object 
 12   NAME_INCOME_TYPE              object 
 13   NAME_EDUCATION_TYPE           object 
 14   NAME_FAMILY_STATUS            object 
 15   NAME_HOUSING_TYPE             object 
 16   REGION_POPULATION_RELATIVE    float64
 17   DAYS_BIRTH                    int64  
 18   DA

The code above indicates the data types present in the DataFrame. Specifically:

- There are 65 columns with 'float64' data type.
- There are 41 columns with 'int64' data type.
- There are 16 columns with 'object' data type.

So, the DataFrame contains a mix of numerical (float and int) and categorical (object) data types.

In [14]:
#Get more detail about dataset continuous variables, which are characterized by 'int' and 'float' data types.
df.select_dtypes(['int','float']).describe()

SK_ID_CURR    TARGET  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
count   307511.00 307511.00     307511.00         307511.00   307511.00   
mean    278180.52      0.08          0.42         168797.92   599026.00   
std     102790.18      0.27          0.72         237123.15   402490.78   
min     100002.00      0.00          0.00          25650.00    45000.00   
25%     189145.50      0.00          0.00         112500.00   270000.00   
50%     278202.00      0.00          0.00         147150.00   513531.00   
75%     367142.50      0.00          1.00         202500.00   808650.00   
max     456255.00      1.00         19.00      117000000.00  4050000.00   

       AMT_ANNUITY  AMT_GOODS_PRICE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
count    307499.00        307233.00                   307511.00   307511.00   
mean      27108.57        538396.21                        0.02   -16037.00   
std       14493.74        369446.46                        0.01     4363.99   
min        1615.50         40500.00                        0.00   -25229.00   
25%       16524.00        238500.00                        0.01   -19682.00   
50%       24903.00        450000.00                        0.02   -15750.00   
75%       34596.00        679500.00                        0.03   -12413.00   
max      258025.50       4050000.00                        0.07    -7489.00   

       DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  \
count      307511.00          307511.00        307511.00    104582.00   
mean        63815.05           -4986.12         -2994.20        12.06   
std        141275.77            3522.89          1509.45        11.94   
min        -17912.00          -24672.00         -7197.00         0.00   
25%         -2760.00           -7479.50         -4299.00         5.00   
50%         -1213.00           -4504.00         -3254.00         9.00   
75%          -289.00           -2010.00         -1720.00        15.00   
max        365243.00               0.00             0.00        91.00   

       FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  \
count   307511.00       307511.00        307511.00         307511.00   
mean         1.00            0.82             0.20              1.00   
std          0.00            0.38             0.40              0.04   
min          0.00            0.00             0.00              0.00   
25%          1.00            1.00             0.00              1.00   
50%          1.00            1.00             0.00              1.00   
75%          1.00            1.00             0.00              1.00   
max          1.00            1.00             1.00              1.00   

       FLAG_PHONE  FLAG_EMAIL  CNT_FAM_MEMBERS  REGION_RATING_CLIENT  \
count   307511.00   307511.00        307509.00             307511.00   
mean         0.28        0.06             2.15                  2.05   
std          0.45        0.23             0.91                  0.51   
min          0.00        0.00             1.00                  1.00   
25%          0.00        0.00             2.00                  2.00   
50%          0.00        0.00             2.00                  2.00   
75%          1.00        0.00             3.00                  2.00   
max          1.00        1.00            20.00                  3.00   

       REGION_RATING_CLIENT_W_CITY  HOUR_APPR_PROCESS_START  \
count                    307511.00                307511.00   
mean                          2.03                    12.06   
std                           0.50                     3.27   
min                           1.00                     0.00   
25%                           2.00                    10.00   
50%                           2.00                    12.00   
75%                           2.00                    14.00   
max                           3.00                    23.00   

       REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  \
count                   307511

### 3. Data Cleaning and Preprocessing
<a class="anchor" id="part-3"></a>


In [15]:
# the sum of the duplicated() function gives the number of rows that appear more once
df.duplicated().sum()

0

There are no duplicate rows in our data.

In [16]:
# Find the number of missing values in each column
df.isna().sum(axis=0)

SK_ID_CURR                           0
TARGET                               0
NAME_CONTRACT_TYPE                   0
CODE_GENDER                          0
FLAG_OWN_CAR                         0
FLAG_OWN_REALTY                      0
CNT_CHILDREN                         0
AMT_INCOME_TOTAL                     0
AMT_CREDIT                           0
AMT_ANNUITY                         12
AMT_GOODS_PRICE                    278
NAME_TYPE_SUITE                   1292
NAME_INCOME_TYPE                     0
NAME_EDUCATION_TYPE                  0
NAME_FAMILY_STATUS                   0
NAME_HOUSING_TYPE                    0
REGION_POPULATION_RELATIVE           0
DAYS_BIRTH                           0
DAYS_EMPLOYED                        0
DAYS_REGISTRATION                    0
DAYS_ID_PUBLISH                      0
OWN_CAR_AGE                     202929
FLAG_MOBIL                           0
FLAG_EMP_PHONE                       0
FLAG_WORK_PHONE                      0
FLAG_CONT_MOBILE         

In [17]:
# Show % null value in each column:
null_percentage = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage

SK_ID_CURR                      0.00
TARGET                          0.00
NAME_CONTRACT_TYPE              0.00
CODE_GENDER                     0.00
FLAG_OWN_CAR                    0.00
FLAG_OWN_REALTY                 0.00
CNT_CHILDREN                    0.00
AMT_INCOME_TOTAL                0.00
AMT_CREDIT                      0.00
AMT_ANNUITY                     0.00
AMT_GOODS_PRICE                 0.09
NAME_TYPE_SUITE                 0.42
NAME_INCOME_TYPE                0.00
NAME_EDUCATION_TYPE             0.00
NAME_FAMILY_STATUS              0.00
NAME_HOUSING_TYPE               0.00
REGION_POPULATION_RELATIVE      0.00
DAYS_BIRTH                      0.00
DAYS_EMPLOYED                   0.00
DAYS_REGISTRATION               0.00
DAYS_ID_PUBLISH                 0.00
OWN_CAR_AGE                    65.99
FLAG_MOBIL                      0.00
FLAG_EMP_PHONE                  0.00
FLAG_WORK_PHONE                 0.00
FLAG_CONT_MOBILE                0.00
FLAG_PHONE                      0.00
F

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns  # Add this line to import seaborn

plt.figure(figsize=(10, 6))
sns.heatmap(df.isnull(), cbar=False, cmap='viridis')
plt.title('Null Values in DataFrame')
plt.show()

# have null values in most of the columns so iam not going to fill null values here i will perfrom some EDA

We will delete any column with over 20% null values, as well as unrelated columns that are not useful to our purpose or whose contents we do not understand. anyutrh]ing ovetr 20% is a large amount of missing value and we dont have insider information so there is no way that we can accurately fill those values thus we deletre those. Before removing columns with over 20% null values or those that seem unrelated or unclear, it's essential to confirm that null values truly indicate missing data and not some other meaningful category, such as 'null' not necessarily implying a binary choice like 'yes' or 'no'. Null values should solely indicate missing data and not carry any additional meaning.

Retrive column names with over 20% null values:

In [ ]:
columns_to_drop = null_percentage[null_percentage > 20]
columns_to_drop

convert the column names with over 20% null values to a list:

In [ ]:
columns_to_drop = columns_to_drop.index.tolist()
columns_to_drop

Add additional unrelated columns to drop:

In [ ]:
unrelated_columns = ['SK_ID_CURR']
unrelated_columns

In [ ]:
# combining both lists of columns to drop
columns_to_drop.extend(unrelated_columns)
columns_to_drop

In [ ]:
df.drop(columns=columns_to_drop, inplace=True)

In [ ]:
df.shape

In [ ]:
# Show % null value in each column:
null_percentage1 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage1

The columns 'EXT_SOURCE_2' and 'EXT_SOURCE_3' represent normalized credit scores from external sources. 'EXT_SOURCE_2' has 0.21% missing data, while 'EXT_SOURCE_3' has 19.83% missing data.To address missing values in the 'EXT_SOURCE_3' column, begin by creating a histogram.

In [ ]:
# Save column values MEAN and MEDIAN to variables
ext_source_3_mean = df['EXT_SOURCE_3'].mean()
ext_source_3_median = df['EXT_SOURCE_3'].median()

# Plot a histogram of EXT_SOURCE_2, with lines marking mean and median
plt.figure(figsize=(10, 6))
plt.hist(df['EXT_SOURCE_3'], bins=50, color='skyblue', edgecolor='black')
plt.axvline(ext_source_3_mean, label='Mean', color='red', linestyle='--')
plt.axvline(ext_source_3_median, label='Median', color='green', linestyle='-')
plt.xlabel('EXT_SOURCE_3')
plt.ylabel('Frequency')
plt.title('Histogram of EXT_SOURCE_3 with Mean and Median')
plt.legend()
plt.show()

In [ ]:
# Calculate skewness
skewness = df['EXT_SOURCE_3'].skew()

# Calculate mean
mean_value = df['EXT_SOURCE_3'].mean()

# Calculate median
median_value = df['EXT_SOURCE_3'].median()

print("Skewness:", skewness)
print("Mean:", mean_value)
print("Median:", median_value)

The negative skewness value (-0.409) suggests that the distribution is slightly left-skewed.

The mean (0.511) is slightly less than the median (0.535), indicating a slight left skewness.
Given the relatively small difference between the mean and median, and the continued slight left skewness, filling null values with either the mean or median is reasonable.

To address missing values in the 'EXT_SOURCE_2' column, begin by creating a histogram.

In [ ]:
# Save column values MEAN and MEDIAN to variables
ext_source_2_mean = df['EXT_SOURCE_2'].mean()
ext_source_2_median = df['EXT_SOURCE_2'].median()

# Plot a histogram of EXT_SOURCE_2, with lines marking mean and median
plt.figure(figsize=(10, 6))
plt.hist(df['EXT_SOURCE_2'], bins=50, color='skyblue', edgecolor='black')
plt.axvline(ext_source_2_mean, label='Mean', color='red', linestyle='--')
plt.axvline(ext_source_2_median, label='Median', color='green', linestyle='-')
plt.xlabel('EXT_SOURCE_2')
plt.ylabel('Frequency')
plt.title('Histogram of EXT_SOURCE_2 with Mean and Median')
plt.legend()
plt.show()


In [ ]:
# Calculate skewness
skewness = df['EXT_SOURCE_2'].skew()

# Calculate mean
mean_value = df['EXT_SOURCE_2'].mean()

# Calculate median
median_value = df['EXT_SOURCE_2'].median()

print("Skewness:", skewness)
print("Mean:", mean_value)
print("Median:", median_value)

The followings are observed from the above plot and the calulation:
The tail on the left side of the distribution is longer than the tail on the right side. The negative skewness value (-0.794) suggests that the distribution of data is still slightly left-skewed, but more so compared to the previous distribution. 
The mean (0.514) is less than the median (0.566),indicating a slight left skewness, although the difference between the mean and median has increased compared to before.
The bulk of the data points are concentrated on the right side of the distribution, while a few extreme values are on the left side.


One approach to address null in the columns 'EXT_SOURCE_2' and 'EXT_SOURCE_3' is to create a new column called 'AVERAGE_CREDIT_SCORE' and then handling the null values in this new column rather than dealing with each of 'EXT_SOURCE_2' and 'EXT_SOURCE_3' separately. This simplifies the data processing and gives a single, consolidated score to work with.

Create the new 'AVERAGE_CREDIT_SCORE' column by calculating the mean of 'EXT_SOURCE_2' and 'EXT_SOURCE_3..

Handle the null values in the new column 'AVERAGE_CREDIT_S and finay d
Drop the original 'EXT_SOURCE_2' and 'EXT_SOURCE_3' columns.

In [ ]:

# Calculate the average credit score, excluding null values
df['AVERAGE_CREDIT_SCORE'] = df[['EXT_SOURCE_2', 'EXT_SOURCE_3']].mean(axis=1, skipna=True)


To handle null values in the new column AVERAGE_CREDIT_SCORE, first visualize the distribution of this column by creating a histogram:

In [ ]:
# Save column values MEAN and MEDIAN to variables
AVERAGE_CREDIT_SCORE_mean = df['AVERAGE_CREDIT_SCORE'].mean()
AVERAGE_CREDIT_SCORE_median = df['AVERAGE_CREDIT_SCORE'].median()

# Plot a histogram of AVERAGE_CREDIT_SCORE, with lines marking mean and median
plt.figure(figsize=(10, 6))
plt.hist(df['AVERAGE_CREDIT_SCORE'], bins=50, color='skyblue', edgecolor='black')
plt.axvline(AVERAGE_CREDIT_SCORE_mean, label='Mean', color='red', linestyle='--')
plt.axvline(AVERAGE_CREDIT_SCORE_median, label='Median', color='green', linestyle='-')
plt.xlabel('AVERAGE_CREDIT_SCORE')
plt.ylabel('Frequency')
plt.title('Histogram of AVERAGE_CREDIT_SCORE with Mean and Median')
plt.legend()
plt.show()

In [ ]:
# Calculate skewness
AVERAGE_CREDIT_SCORE_skewness = df['AVERAGE_CREDIT_SCORE'].skew()

# Calculate mean
AVERAGE_CREDIT_SCORE_mean_value = df['AVERAGE_CREDIT_SCORE'].mean()

# Calculate median
AVERAGE_CREDIT_SCORE_median_value = df['AVERAGE_CREDIT_SCORE'].median()

print("Skewness:", AVERAGE_CREDIT_SCORE_skewness)
print("Mean:", AVERAGE_CREDIT_SCORE_mean_value)
print("Median:", AVERAGE_CREDIT_SCORE_median_value)

The negative skewness value (-0.623) suggests that the distribution of data is slightly left-skewed, meaning it has a longer tail on the left side. 

The mean (0.511) is slightly less than the median (0.532), indicating that the distribution is negatively skewed, as expected.

Overall,even though there is a slight left skewness, the distribution is relatively close to symmetric, with the median and mean not significantly different.

Given that the distribution is slightly left-skewed and the mean and median are close, either the mean or median could be suitable for filling null values. However, since the mean is slightly lower than the median in this case, using the median might be a more conservative approach to ensure the filled values align with the central tendency of the data. 

In [ ]:
# Calculate the median of the 'AVERAGE_CREDIT_SCORE' column
median_credit_score = df['AVERAGE_CREDIT_SCORE'].median()

# Fill null values with the median
df['AVERAGE_CREDIT_SCORE'].fillna(median_credit_score, inplace=True)


# Drop the original columns
df.drop(['EXT_SOURCE_2', 'EXT_SOURCE_3'], axis=1, inplace=True)


In [ ]:
# Show % null value in each column to confirm the changes were implemented:
null_percentage2 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage2

In [ ]:
df.head()

To find where the majority of loans fall, begin by generating a description of the 'AMT_CREDIT' column using the `describe()` method.

In [ ]:
# Get the summary description of the 'AMT_CREDIT' column
amt_credit_description = df['AMT_CREDIT'].describe()
amt_credit_description

Then create a histogram plot for the AMT_CREDIT column:

In [ ]:
# convert infinite values to NaN before plotting
df.replace([np.inf, -np.inf], np.nan, inplace=True)

plt.figure(figsize=(10, 8))

# Plot using Seaborn
sns.histplot(df['AMT_CREDIT'], bins=50, kde=False)

# Calculate quartiles and median
Q1 = df['AMT_CREDIT'].quantile(0.25)
median = df['AMT_CREDIT'].median()
Q3 = df['AMT_CREDIT'].quantile(0.75)

# Plot quartiles and median as vertical lines
plt.axvline(x=Q1, color='red', linestyle='--', label='Q1')
plt.axvline(x=median, color='green', linestyle='--', label='Median')
plt.axvline(x=Q3, color='blue', linestyle='--', label='Q3')

# Set the ticker format for the x-axis to display actual loan amounts
plt.ticklabel_format(style='plain', axis='x')

# Add labels, title, and legend for better understanding
plt.xlabel('Loan Amount')
plt.ylabel('Frequency')
plt.title('Distribution of Loan Amounts')
plt.legend()

# Print quartiles and median
print("Q1 (25th Percentile):", Q1)
print("Median (50th Percentile):", median)
print("Q3 (75th Percentile):", Q3)

# Show the plot
plt.show()


The median loan amount is 513,531, meaning half of the loans are below this amount and indicates the central tendency of the data.
- (0.25): 270,000 - This means that 25% of the loan amounts are less than or equal to 270,000.
- 50th Percentile (Median, 0.50): 513,531 - This indicates that 50% of the loan amounts are less than or equal to 513,531.
-  75th Percentile (0.75): 808,650 - This tells us that 75% of the loan amounts are less than or equal to 808,650.

Filter the data based on the majority that falls within a certain range by using the interquartile range (IQR) method. 
Filter out data points that fall outside a specified range defined by the first quartile (Q1) minus 1.5 times the IQR and the third quartile (Q3) plus 1.5 times the IQR. 





In [ ]:
# Calculate quartiles
Q1 = df['AMT_CREDIT'].quantile(0.25)
Q3 = df['AMT_CREDIT'].quantile(0.75)

# Calculate the interquartile range (IQR)
IQR = Q3 - Q1

# Define the lower and upper bounds for filtering the majority
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter the majority of data within the specified range
#the original DataFrame df will be overwritten  with the filtered data.
df = df[(df['AMT_CREDIT'] >= lower_bound) & (df['AMT_CREDIT'] <= upper_bound)]


In [ ]:
df.shape

In [ ]:
# Show % null value in each column:
null_percentage4 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage4

The following columns have 13.58% missing values:

AMT_REQ_CREDIT_BUREAU_HOUR
AMT_REQ_CREDIT_BUREAU_DAY
AMT_REQ_CREDIT_BUREAU_WEEK
AMT_REQ_CREDIT_BUREAU_MON
AMT_REQ_CREDIT_BUREAU_QRT
AMT_REQ_CREDIT_BUREAU_YEAR
AVERAGE_CREDIT_SCORE
Addressing Null Values:
To handle the null values in these columns, begin by finding the minimum and maximum values for each column.

In [ ]:
# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_HOUR
min_amt_req_credit_bureau_hour = df['AMT_REQ_CREDIT_BUREAU_HOUR'].min()
max_amt_req_credit_bureau_hour = df['AMT_REQ_CREDIT_BUREAU_HOUR'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_HOUR:", min_amt_req_credit_bureau_hour)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_HOUR:", max_amt_req_credit_bureau_hour)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_DAY
min_amt_req_credit_bureau_day = df['AMT_REQ_CREDIT_BUREAU_DAY'].min()
max_amt_req_credit_bureau_day = df['AMT_REQ_CREDIT_BUREAU_DAY'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_DAY:", min_amt_req_credit_bureau_day)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_DAY:", max_amt_req_credit_bureau_day)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_WEEK
min_amt_req_credit_bureau_week = df['AMT_REQ_CREDIT_BUREAU_WEEK'].min()
max_amt_req_credit_bureau_week = df['AMT_REQ_CREDIT_BUREAU_WEEK'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_WEEK:", min_amt_req_credit_bureau_week)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_WEEK:", max_amt_req_credit_bureau_week)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_MON
min_amt_req_credit_bureau_mon = df['AMT_REQ_CREDIT_BUREAU_MON'].min()
max_amt_req_credit_bureau_mon = df['AMT_REQ_CREDIT_BUREAU_MON'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_MON:", min_amt_req_credit_bureau_mon)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_MON:", max_amt_req_credit_bureau_mon)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_QRT
min_amt_req_credit_bureau_qrt = df['AMT_REQ_CREDIT_BUREAU_QRT'].min()
max_amt_req_credit_bureau_qrt = df['AMT_REQ_CREDIT_BUREAU_QRT'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_QRT:", min_amt_req_credit_bureau_qrt)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_QRT:", max_amt_req_credit_bureau_qrt)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_YEAR
min_amt_req_credit_bureau_year = df['AMT_REQ_CREDIT_BUREAU_YEAR'].min()
max_amt_req_credit_bureau_year = df['AMT_REQ_CREDIT_BUREAU_YEAR'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_YEAR:", min_amt_req_credit_bureau_year)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_YEAR:", max_amt_req_credit_bureau_year)




The maximum value for 'AMT_REQ_CREDIT_BUREAU_QRT'  (261.0) appears to be an outlier or incorrect. Given that 261.0 cannot be reasonably explained, it is best to remove this value altogether from the dataset to ensure data reliability.

In [ ]:
# Drop the row where the value 261 appears only in the 'AMT_REQ_CREDIT_BUREAU_QRT' column
df = df.drop(df[df['AMT_REQ_CREDIT_BUREAU_QRT'] == 261.0].index)


In [ ]:
# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_HOUR
min_amt_req_credit_bureau_hour = df['AMT_REQ_CREDIT_BUREAU_HOUR'].min()
max_amt_req_credit_bureau_hour = df['AMT_REQ_CREDIT_BUREAU_HOUR'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_HOUR:", min_amt_req_credit_bureau_hour)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_HOUR:", max_amt_req_credit_bureau_hour)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_DAY
min_amt_req_credit_bureau_day = df['AMT_REQ_CREDIT_BUREAU_DAY'].min()
max_amt_req_credit_bureau_day = df['AMT_REQ_CREDIT_BUREAU_DAY'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_DAY:", min_amt_req_credit_bureau_day)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_DAY:", max_amt_req_credit_bureau_day)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_WEEK
min_amt_req_credit_bureau_week = df['AMT_REQ_CREDIT_BUREAU_WEEK'].min()
max_amt_req_credit_bureau_week = df['AMT_REQ_CREDIT_BUREAU_WEEK'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_WEEK:", min_amt_req_credit_bureau_week)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_WEEK:", max_amt_req_credit_bureau_week)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_MON
min_amt_req_credit_bureau_mon = df['AMT_REQ_CREDIT_BUREAU_MON'].min()
max_amt_req_credit_bureau_mon = df['AMT_REQ_CREDIT_BUREAU_MON'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_MON:", min_amt_req_credit_bureau_mon)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_MON:", max_amt_req_credit_bureau_mon)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_QRT1
min_amt_req_credit_bureau_qrt1 = df['AMT_REQ_CREDIT_BUREAU_QRT'].min()
max_amt_req_credit_bureau_qrt1 = df['AMT_REQ_CREDIT_BUREAU_QRT'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_QRT1:", min_amt_req_credit_bureau_qrt)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_QRT1:", max_amt_req_credit_bureau_qrt)


# Find the minimum and maximum values for AMT_REQ_CREDIT_BUREAU_YEAR
min_amt_req_credit_bureau_year = df['AMT_REQ_CREDIT_BUREAU_YEAR'].min()
max_amt_req_credit_bureau_year = df['AMT_REQ_CREDIT_BUREAU_YEAR'].max()
print("Minimum value for AMT_REQ_CREDIT_BUREAU_YEAR:", min_amt_req_credit_bureau_year)
print("Maximum value for AMT_REQ_CREDIT_BUREAU_YEAR:", max_amt_req_credit_bureau_year)



In [ ]:
#Let's confirm that the number of rows has been reduced by one after removing the outlier value from the 'AMT_REQ_CREDIT_BUREAU_QRT' column.
df.shape

In [ ]:
# Show % null value in each column:
null_percentage5 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage5

To address the 13.58% missing values in the columns AMT_REQ_CREDIT_BUREAU_HOUR, AMT_REQ_CREDIT_BUREAU_DAY, AMT_REQ_CREDIT_BUREAU_WEEK, AMT_REQ_CREDIT_BUREAU_MON, AMT_REQ_CREDIT_BUREAU_QRT, and AMT_REQ_CREDIT_BUREAU_YEAR, assign Weights to Each Column based on their importance. The weights are as follows:
- 'AMT_REQ_CREDIT_BUREAU_HOUR': 6
- 'AMT_REQ_CREDIT_BUREAU_DAY': 5 5
- 'AMT_REQ_CREDIT_BUREAU_WEEK': 4
- 'AMT_REQ_CREDIT_BUREAU_MON': 3
- 'AMT_REQ_CREDIT_BUREAU_QRT': 2
- 'AMT_REQ_CREDIT_BUREAU_Y
- R': eCalculate the weighted sum for each row by multiplying the value in each column by its corresponding w.set."

In [ ]:
# Calculate the weighted sum
weighted_sum = (df['AMT_REQ_CREDIT_BUREAU_HOUR'] * 6 +
                df['AMT_REQ_CREDIT_BUREAU_DAY'] * 5 +
                df['AMT_REQ_CREDIT_BUREAU_WEEK'] * 4 +
                df['AMT_REQ_CREDIT_BUREAU_MON'] * 3 +
                df['AMT_REQ_CREDIT_BUREAU_QRT'] * 2 +
                df['AMT_REQ_CREDIT_BUREAU_YEAR'] * 1)


In [ ]:
# Create a new column with the weighted sum
df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'] = weighted_sum

# Drop the original columns
df = df.drop(columns=['AMT_REQ_CREDIT_BUREAU_HOUR',
                      'AMT_REQ_CREDIT_BUREAU_DAY',
                      'AMT_REQ_CREDIT_BUREAU_WEEK', 
                      'AMT_REQ_CREDIT_BUREAU_MON', 
                      'AMT_REQ_CREDIT_BUREAU_QRT', 
                      'AMT_REQ_CREDIT_BUREAU_YEAR'])

In [ ]:
# Confirm that we have a new column WEIGHTED_SUM_REQ_CREDIT_BUREAU.
df.head()

We have a new column WEIGHTED_SUM_REQ_CREDIT_BUREAU that captures the combined information from the original columns

In [ ]:
# Show % null value in each column:
null_percentage6 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage6

After calculating the weighted sum, the new column WEIGHTED_SUM_REQ_CREDIT_BUREAU also has 13.58% null values, as the original columns had. To handle these missing values, let's draw a histogram showing the mean and median to understand the distribution of the data.

In [ ]:
# Save column values MEAN, MEDIAN, and MODE to variables
WEIGHTED_SUM_REQ_CREDIT_BUREAU_mean = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].mean()
WEIGHTED_SUM_REQ_CREDIT_BUREAU_median = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].median()
WEIGHTED_SUM_REQ_CREDIT_BUREAU_mode = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].mode()[0]

# Plot a histogram of WEIGHTED_SUM_REQ_CREDIT_BUREAU, with lines marking mean, median, and mode
plt.figure(figsize=(10, 6))
plt.hist(df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'], bins=50, color='skyblue', edgecolor='black')

# Plot mean, median, and mode lines
plt.axvline(WEIGHTED_SUM_REQ_CREDIT_BUREAU_mean, label=f'Mean: {WEIGHTED_SUM_REQ_CREDIT_BUREAU_mean:.2f}', color='red', linestyle='--', linewidth=2)
plt.axvline(WEIGHTED_SUM_REQ_CREDIT_BUREAU_median, label=f'Median: {WEIGHTED_SUM_REQ_CREDIT_BUREAU_median:.2f}', color='green', linestyle='-', linewidth=2)
plt.axvline(WEIGHTED_SUM_REQ_CREDIT_BUREAU_mode, label=f'Mode: {WEIGHTED_SUM_REQ_CREDIT_BUREAU_mode:.2f}', color='blue', linestyle=':', linewidth=2)

# Add text annotations for mean, median, and mode
plt.text(WEIGHTED_SUM_REQ_CREDIT_BUREAU_mean, plt.ylim()[1] * 0.9, f'Mean: {WEIGHTED_SUM_REQ_CREDIT_BUREAU_mean:.2f}', color='red', ha='center')
plt.text(WEIGHTED_SUM_REQ_CREDIT_BUREAU_median, plt.ylim()[1] * 0.8, f'Median: {WEIGHTED_SUM_REQ_CREDIT_BUREAU_median:.2f}', color='green', ha='center')
plt.text(WEIGHTED_SUM_REQ_CREDIT_BUREAU_mode, plt.ylim()[1] * 0.7, f'Mode: {WEIGHTED_SUM_REQ_CREDIT_BUREAU_mode:.2f}', color='blue', ha='center')

# Add labels and title for the plot
plt.xlabel('WEIGHTED_SUM_REQ_CREDIT_BUREAU')
plt.ylabel('Frequency')
plt.title('Histogram of WEIGHTED_SUM_REQ_CREDIT_BUREAU with Mean, Median, and Mode')
plt.legend()

# Show the plot
plt.show()



In [ ]:
# Calculate skewness
skewness = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].skew()

# Calculate mean
mean_value = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].mean()

# Calculate median
median_value = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].median()

# Calculate mode
mode_value = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].mode()

print("Skewness:", skewness)
print("Mean:", mean_value)
print("Median:", median_value)
print("Mode:", mode_value)

The high positive skewness (3.64) indicates that the distribution is right-skewed, meaning that the tail of the distribution extends to the right, with a majority of values concentrated on the left side.
Given the skewness and the presence of outliers, use the median to fill null values in this column as the median is less sensitive to outliers and provides a better representation of the central tendency in skewed distributions.

In [ ]:
# Calculate the median of the WEIGHTED_SUM_REQ_CREDIT_BUREAU column
median_value = df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].median()

# Fill the missing values in the WEIGHTED_SUM_REQ_CREDIT_BUREAU column with the median value
df['WEIGHTED_SUM_REQ_CREDIT_BUREAU'].fillna(median_value, inplace=True)


In [ ]:
# Confirm the missing values are filled
null_percentage7 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage7

After handling the missing values in the WEIGHTED_SUM_REQ_CREDIT_BUREAU column, the column no longer contains any missing values, indicating that all null values have been filled.

Let's create histograms for each of the columns: `OBS_30_CNT_SOCIAL_CIRCLE`, `DEF_30_CNT_SOCIAL_CIRCLE`, `OBS_60_CNT_SOCIAL_CIRCLE`, and `DEF_60_CNT_SOCIAL_CIRCLE` to address the 0.34% null values.

In [ ]:
# Set the figure size
plt.figure(figsize=(20, 15))

# Plot histogram for OBS_30_CNT_SOCIAL_CIRCLE
plt.subplot(2, 2, 1)
sns.histplot(df['OBS_30_CNT_SOCIAL_CIRCLE'], bins=30, kde=False, color='skyblue')
plt.title('OBS_30_CNT_SOCIAL_CIRCLE')
plt.xlabel('OBS_30_CNT_SOCIAL_CIRCLE')
plt.ylabel('Frequency')

# Plot histogram for DEF_30_CNT_SOCIAL_CIRCLE
plt.subplot(2, 2, 2)
sns.histplot(df['DEF_30_CNT_SOCIAL_CIRCLE'], bins=30, kde=False, color='skyblue')
plt.title('DEF_30_CNT_SOCIAL_CIRCLE')
plt.xlabel('DEF_30_CNT_SOCIAL_CIRCLE')
plt.ylabel('Frequency')

# Plot histogram for OBS_60_CNT_SOCIAL_CIRCLE
plt.subplot(2, 2, 3)
sns.histplot(df['OBS_60_CNT_SOCIAL_CIRCLE'], bins=30, kde=False, color='skyblue')
plt.title('OBS_60_CNT_SOCIAL_CIRCLE')
plt.xlabel('OBS_60_CNT_SOCIAL_CIRCLE')
plt.ylabel('Frequency')

# Plot histogram for DEF_60_CNT_SOCIAL_CIRCLE
plt.subplot(2, 2, 4)
sns.histplot(df['DEF_60_CNT_SOCIAL_CIRCLE'], bins=30, kde=False, color='skyblue')
plt.title('DEF_60_CNT_SOCIAL_CIRCLE')
plt.xlabel('DEF_60_CNT_SOCIAL_CIRCLE')
plt.ylabel('Frequency')

# Show the plots
plt.tight_layout()
plt.show()

The above columns are heavily right-skewed, with most values concentrated around zero. Given this distribution, filling null values with zero (0) is a reasonable approach. However, since we have highly skewed distribution and the presence of many zero values, filling null values with the median is a better approach as median is less sensitive to extreme values and outliers and represents the central tendency in a skewed distribution. 

In [ ]:
# Impute missing values with the mode for each column
df['OBS_30_CNT_SOCIAL_CIRCLE'].fillna(df['OBS_30_CNT_SOCIAL_CIRCLE'].median(), inplace=True)
df['DEF_30_CNT_SOCIAL_CIRCLE'].fillna(df['DEF_30_CNT_SOCIAL_CIRCLE'].median(), inplace=True)
df['OBS_60_CNT_SOCIAL_CIRCLE'].fillna(df['OBS_60_CNT_SOCIAL_CIRCLE'].median(), inplace=True)
df['DEF_60_CNT_SOCIAL_CIRCLE'].fillna(df['DEF_60_CNT_SOCIAL_CIRCLE'].median(), inplace=True)


In [ ]:
# Confirm the missing values are filled
null_percentage8 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage8

AMT_GOODS_PRICE has 0.09% null values. Check the distribution of the column AMT_GOODS_PRICE:

In [ ]:

# Calculate mean, median, and mode
AMT_GOODS_PRICE_mean = df['AMT_GOODS_PRICE'].mean()
AMT_GOODS_PRICE_median = df['AMT_GOODS_PRICE'].median()
AMT_GOODS_PRICE_mode = df['AMT_GOODS_PRICE'].mode()[0]

# Plot a histogram of AMT_GOODS_PRICE, with lines marking mean, median, and mode
plt.figure(figsize=(10, 6))
plt.hist(df['AMT_GOODS_PRICE'], bins=50, color='skyblue', edgecolor='black')

# Plot mean, median, and mode lines
plt.axvline(AMT_GOODS_PRICE_mean, label=f'Mean: {AMT_GOODS_PRICE_mean:.0f}', color='red', linestyle='--', linewidth=2)
plt.axvline(AMT_GOODS_PRICE_median, label=f'Median: {AMT_GOODS_PRICE_median:.0f}', color='green', linestyle='-', linewidth=2)
plt.axvline(AMT_GOODS_PRICE_mode, label=f'Mode: {AMT_GOODS_PRICE_mode:.0f}', color='blue', linestyle=':', linewidth=2)

# Add text annotations for mean, median, and mode
plt.text(AMT_GOODS_PRICE_mean, plt.ylim()[1] * 0.9, f'Mean: {AMT_GOODS_PRICE_mean:.0f}', color='red', ha='center')
plt.text(AMT_GOODS_PRICE_median, plt.ylim()[1] * 0.8, f'Median: {AMT_GOODS_PRICE_median:.0f}', color='green', ha='center')
plt.text(AMT_GOODS_PRICE_mode, plt.ylim()[1] * 0.7, f'Mode: {AMT_GOODS_PRICE_mode:.0f}', color='blue', ha='center')

# Add labels and title for the plot
plt.xlabel('AMT_GOODS_PRICE')
plt.ylabel('Frequency')
plt.title('Histogram of AMT_GOODS_PRICE with Mean, Median, and Mode')
plt.legend()

# Show the plot
plt.show()



In [ ]:
# Calculate skewness
skewness = df['AMT_GOODS_PRICE'].skew()

# Calculate mean
mean_value = df['AMT_GOODS_PRICE'].mean()

# Calculate median
median_value = df['AMT_GOODS_PRICE'].median()

# Calculate mode
mode_value = df['AMT_GOODS_PRICE'].mode()

print("Skewness:", skewness)
print("Mean:", mean_value)
print("Median:", median_value)
print("Mode:", mode_value)

The distribution of 'AMT_GOODS_PRICE' is right-skewed, with a majority of values clustered around the mode and some higher values pulling the mean to the right..

In [ ]:
# Impute missing values with the median for AMT_GOODS_PRICE' column
df['AMT_GOODS_PRICE'].fillna(df['AMT_GOODS_PRICE'].median, inplace=True)

In [ ]:
# Confirm the missing values are filled
null_percentage9 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage9

NAME_TYPE_SUITE column has 0.40% missing data. Check the Distribution of 'NAME_TYPE_SUITE'.

In [ ]:
# Plot a Bar Chart to Visualize the Distribution
plt.figure(figsize=(12, 6))
sns.histplot(df['NAME_TYPE_SUITE'], color='skyblue', edgecolor='black', discrete=True)
plt.xlabel('NAME_TYPE_SUITE')
plt.ylabel('Frequency')
plt.title('Distribution of NAME_TYPE_SUITE (Before Dropping Null)')

# Show the plot
plt.show()


In [ ]:
# Display descriptive statistics
name_type_suite_mode = df['NAME_TYPE_SUITE'].mode()[0]
print(f"Mode of NAME_TYPE_SUITE: {name_type_suite_mode}")


The above results show that the majority of applicants were unaccompanied when applying for a loan. Given that the mode represents the most common value in the column, replace null values with the mode.

In [ ]:
# Impute missing values with the mode for each column
df['NAME_TYPE_SUITE'].fillna(df['NAME_TYPE_SUITE'].mode()[0], inplace=True)


In [ ]:
# Confirm the missing values are filled
null_percentage10 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage10

In [ ]:
df.shape

We have multiple columns, like FLAG_DOCUMENT_2, FLAG_DOCUMENT_3, etc., each indicating whether the client provided a specific document.
Instead of having too many columns, create a single column that captures the total count of documents provided by the applicant.

In [ ]:
# Check unique values in the document columns
unique_values = df[['FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5',
                    'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9',
                    'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13',
                    'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17',
                    'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']].stack().unique()

print(unique_values)


In [ ]:
# Create a new column 'DOCUMENT_COUNT' which indicates the total number of documents provided
df['DOCUMENT_COUNT'] = df[['FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5',
                            'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9',
                            'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13',
                            'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17',
                            'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']].sum(axis=1)

In [ ]:
df.head()

In [ ]:
df.shape

The number of columns increased from 65 to 66, indicating that the column DOCUMENT_COUNT was successfully generated. Now, drop all the original columns after creating the 'DOCUMENT_COUNT' column.

In [ ]:
# Drop the original columns without creating a list
df.drop(columns=['FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5',
                 'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9',
                 'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13',
                 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17',
                 'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21'], inplace=True)

In [ ]:
df.shape

In [ ]:
# Print unique values in the 'DOCUMENT_COUNT' column
unique_values = df['DOCUMENT_COUNT'].unique()
print("Unique values in DOCUMENT_COUNT column:", unique_values)

# Print value counts for each unique value
value_counts = df['DOCUMENT_COUNT'].value_counts()
print("\nValue counts for DOCUMENT_COUNT column:")
print(value_counts)


In [ ]:
df.head()

In [ ]:
# check the remaining column and the missing values:
null_percentage11 = round(df.isnull().sum() / df.shape[0] * 100.00,2)
null_percentage11

In [ ]:
df.shape

### 4. Exploratory Data Analysis (EDA)
<a class="anchor" id="part-4"></a>


After completing the data cleaning process, our dataset now contains 300,948 rows and 65 columns. With the missing values addressed and irrelevant columns dropped, we are ready to proceed to the next crucial phase: Exploratory Data Analysis (EDA). Now let's do some analysis.


In [ ]:
# Plot the distribution of our interestvariable: TARGET
df["TARGET"].value_counts(normalize=True).plot(kind="bar")
plt.show()

one representes instances where a default occurs and the zero representes instances where it does not.

In [ ]:
# Get the count of each unique value in 'loan_status' column
df['TARGET'].value_counts()

In [ ]:
# Calculate the counts of each class in the target variable
target_counts = df['TARGET'].value_counts()

# Define labels for the pie chart
labels = ['Non-Defaulters', 'Defaulters']

# Create a pie chart
plt.figure(figsize=(8, 8))
plt.pie(target_counts, labels=labels, autopct='%1.1f%%', startangle=140, colors=['#66b3ff','#ff6666'])

# Add a title
plt.title('Distribution of Loan Defaulters vs. Non-Defaulters')

plt.show()
print()
print('We can observe from the plot that 91.8% of customers are paying their loan amounts, while 8.2% are going into default')
print()

The above distribution suggests that the majority of instances do not result in a default, as the percentage of instances with no default is significantly higher than those with a default. 

In [ ]:
# Plotting the distribution of income by loan status and gender
plt.figure(figsize=(12, 6))
sns.barplot(data=df, x='TARGET', y='AMT_INCOME_TOTAL', hue='CODE_GENDER', ci=None, palette='coolwarm')
plt.title('Distribution of Income by Loan Status and Gender')
plt.xlabel('Loan Status')
plt.ylabel('Income')
plt.show()

# Calculate summary statistics for income by loan status and gender
income_summary = df.groupby(['TARGET', 'CODE_GENDER'])['AMT_INCOME_TOTAL'].agg(['mean', 'median', 'std', 'min', 'max', 'count']).reset_index()

# Display the summary table
print(income_summary)


Males have higher median incomes than females, and the incomes for non-defaulters are higher than those for defaulters among males.
Non-defaulter males have higher mean and median incomes compared to non-defaulter females.
Defaulting males also have higher mean and median incomes compared to defaulting females.

In [ ]:
# Convert 'TARGET' column to numeric
df['TARGET'] = pd.to_numeric(df['TARGET'], errors='coerce')

# Group by education level and calculate the default rate
education_default_rate = df.groupby('NAME_EDUCATION_TYPE')['TARGET'].mean().reset_index()

# Rename the columns for better understanding
education_default_rate.columns = ['Education Level', 'Default Rate']

print(education_default_rate)


In [ ]:
# Plot the default rates for each education level
plt.figure(figsize=(10, 6))
sns.barplot(data=education_default_rate, x='Default Rate', y='Education Level', palette='viridis')
plt.title('Default Rate by Education Level')
plt.xlabel('Default Rate')
plt.ylabel('Education Level')
plt.show()



Academic degeer categoryry has the lowest default rate at 0.02, indicating that clients with academic degrees tend to default less oftenclients with .

Higher educat haee a relatively low default rate of 0.05, suggesting that a higher level of education may be associated with lower defaulrates.

L lower secondary and secondary/secondary special education backgrounds have the highest default rates at 0.11 and 0.09, respectively. This suggests a tendency for higher default occurrences among individuals with lower educationat leve.n.

#  Exploring correlations between variables¶

One final kind of relationship to always look at is the correlation between the different variables in  dataset. Not only may this reveal some interesting relationships within  dataset; it's also important for modelling purposes, as variables that are highly correlated with one another can lead to spurious results if both are included in your model.

The easiest way to explore these correlations is with a correlation matrix:

In [ ]:
# Identify and exclude non-numeric columns
numeric_columns = df.select_dtypes(include=['number']).columns
numeric_df = df[numeric_columns]

# Now, compute the correlation matrix
correlation_matrix = numeric_df.corr()

# Display the correlation matrix
correlation_matrix


This dataframe displays the correlation between all the variables in our dataset; however, it is not at all easy to read at a glance. It's much better to visualize it as a heatmap as visualizing the correlation matrix as a heatmap makes it easier to understand and interpret the relationships between variables. 

In [ ]:
# Select only numeric columns
numeric_df = df.select_dtypes(include=['number'])

# Plotting correlation heatmap
plt.figure(figsize=(12, 8))
sns.heatmap(numeric_df.corr(), annot=True, cmap='YlGnBu', cbar=True, square=True, linewidths=.5)
plt.show()


To determine which columns are highly correlated and potentially redundant, we can look for pairs of columns with correlation coefficients close to 1 or -1. Here are some pairs of columns with high correlation coefficients:

CNT_FAM_MEMBERS and CNT_CHILDREN: These columns have a correlation coefficient of approximately 0.88, indicating a strong positive correlation. Since they both represent family size or number of dependents, one of them can be dropped to reduce redundancy.

In [ ]:
df.drop(columns=['CNT_CHILDREN'], inplace=True)

REGION_RATING_CLIENT and REGION_RATING_CLIENT_W_CITY: These columns have correlation coefficients close to 1  indicating a strong positive correlation. They both represent the client's rating of the region, so one of them could be dropped.

In [ ]:
df.drop(columns=['REGION_RATING_CLIENT_W_CITY'], inplace=True)

OBS_30_CNT_SOCIAL_CIRCLE, DEF_30_CNT_SOCIAL_CIRCLE, OBS_60_CNT_SOCIAL_CIRCLE, and DEF_60_CNT_SOCIAL_CIRCLE: These columns have high correlation coefficients among themselves, indicating redundancy. They all relate to the client's social circle observations, so drop 3 of them.

In [ ]:
df.drop(columns=['OBS_30_CNT_SOCIAL_CIRCLE', 'DEF_30_CNT_SOCIAL_CIRCLE', 'OBS_60_CNT_SOCIAL_CIRCLE'], inplace=True)

DAYS_BIRTH and AGE represent the same information (age), so we drop one of them.

The column DAYS_BIRTH has negative values that represent the number of days since the client was born. These values are represented as negative integers because they count backward from a reference date. Drop DAYS_BIRTH column as the values are not interpretable for analysis.  

In [ ]:
df.drop(columns=['DAYS_BIRTH'], inplace=True)

LIVE_CITY_NOT_WORK_CITY and REG_CITY_NOT_WORK_CITY have a strong positive correlation. They both  represent whether the client lives and works in the same city, so one of them could be removed.

In [ ]:
df.drop(columns=['REG_CITY_NOT_WORK_CITY'], inplace=True)

Drop columns AMT_CREDIT and AMT_ANNUITY as there could be bias introduced by including them, especially if they are affected by factors that are also related to the target variable (default status).

In [ ]:
df.drop(columns=['AMT_CREDIT', 'AMT_ANNUITY'], inplace=True)

In [ ]:
df.shape

The successful removal of redundant columns has resulted in a reduced dataset size.The number of rows have been reduced from 47 to 37. 

In [ ]:
# Select only numeric columns
numeric_df = df.select_dtypes(include=['number'])

# Plotting correlation heatmap
plt.figure(figsize=(12, 8))
sns.heatmap(numeric_df.corr(), annot=True, cmap='YlGnBu', cbar=True, square=True, linewidths=.5)
plt.show()

Since "REG_REGION_NOT_LIVE_REGION" and "REG_REGION_NOT_WORK_REGION" are highly correlated, drop one of them to avoid multicollinearity issues.

In [ ]:
df.drop(columns=['REG_REGION_NOT_WORK_REGION'], inplace=True)

In [ ]:
df.shape

'FLAG_EMP_PHONE' and DAYS_EMPLOYED' are highly correlated so we keep only one column 

In [ ]:
df.info(verbose=True)

In [ ]:
df.drop(columns=['FLAG_EMP_PHONE'], inplace=True)

In [ ]:
df.shape

In [ ]:
# Diplay the unique values of addr_state
df['DAYS_ID_PUBLISH'].nunique()

Since there are too many unique values to convert to a dummy variable feature. Let's remove 'DAYS_ID_PUBLISH'

In [ ]:
# Drop the emp_title column
df = df.drop('DAYS_ID_PUBLISH',axis=1)

In [ ]:
# Filter to get only the numerical columns needed
numerical_columns = ['TARGET', 'AMT_INCOME_TOTAL', 'REGION_POPULATION_RELATIVE', 
                     'DAYS_EMPLOYED', 'DAYS_REGISTRATION', 'FLAG_MOBIL', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE', 
                     'FLAG_PHONE', 'FLAG_EMAIL', 'CNT_FAM_MEMBERS', 
                     'REGION_RATING_CLIENT', 'HOUR_APPR_PROCESS_START', 
                     'REG_REGION_NOT_LIVE_REGION', 'LIVE_REGION_NOT_WORK_REGION', 
                     'REG_CITY_NOT_LIVE_CITY', 'LIVE_CITY_NOT_WORK_CITY', 
                     'DEF_60_CNT_SOCIAL_CIRCLE', 'DAYS_LAST_PHONE_CHANGE', 
                     'AVERAGE_CREDIT_SCORE', 'WEIGHTED_SUM_REQ_CREDIT_BUREAU', 
                     'DOCUMENT_COUNT']

# Ensure only the specified columns are selected
numeric_df = df[numerical_columns].select_dtypes(include=[np.number])

# Create a mask for the upper-triangular part of the correlation matrix
mask = np.triu(numeric_df.corr())

# Generate a heatmap of the correlation matrix
plt.figure(figsize=(20, 20))
heatmap = sns.heatmap(numeric_df.corr().round(2), annot=True, annot_kws={"size": 10}, vmax=1, vmin=-1, cmap='YlGnBu', mask=mask, center=0, cbar_kws={"shrink": .8},cbar=True, square=True, linewidths=.5)
plt.xticks(rotation=45, ha='right', fontsize=10)
plt.yticks(rotation=0, fontsize=10)
plt.tight_layout()  # Adjust layout to make room for annotations
plt.show()



In [ ]:
df.shape

Currently, the cleaned data has been reduced to 34 columns and 300,948 rows  which will be carried forward to the next project phase.

### 5. Next Steps: Data Processing, Feature Engineering, and Baseline Modeling
<a class="anchor" id="part-5"></a>


In the future, before proceeding with modeling, I will focus on feature engineering by creating binary columns for all categorical variables and conducting hypothesis testing. The null hypothesis (H0) for our linear regression model states that the coefficients of all independent variables are equal to zero, indicating that none of the variables have an effect on loan default. Conversely, the alternative hypothesis (H1) indicates that at least one coefficient is not equal to zero, suggesting that at least one independent variable does influence loan default.

To further prepare the data for this analysis, I will first convert all categorical columns into numerical data using binary encoding (get_dummies C converting these columnsl enabsle the inclusion of all relevant variables in thr regression modelThe r goal is to reduce the dataset to around 20 significant features, utilizing techniqes such as correlation analys.s. This refined dataset will facilitate a more efficient and accurate model training process.